tsz-compress/src/v2/queue.rs

```diff
@@ -0,0 +1,281 @@
+        }
+    }
+
+    ///
+    /// Pops the oldest value from the queue,
+    /// returning None if the queue is empty.
+    ///
+    pub fn pop(&mut self) -> Option<usize> {
+        if self.is_empty() {
+            return None;
+        }
+
+        let value = unsafe { self.value_at(self.front) };
+        self.front = (self.front + 1) % 16;
+        self.len -= 1;
+        Some(value)
+    }
+
+    ///
+    /// Pop N values from the queue at once,
+    /// values may not be meaningful if the queue is
+    /// not of length N.
+    ///
+    #[inline(always)]
+    pub fn pop_n<const M: usize>(&mut self) -> [usize; M] {
+        let mut values: [usize; M] = [0; M];
+        for i in 0..M {
+            let index = (self.front + i) % 16;
+            unsafe {
+                *values.get_unchecked_mut(i) = self.value_at(index);
+            }
+        }
+        self.front = (self.front + M) % 16;
+        self.len -= M;
+        values
+    }
+
+    ///
+    /// Peak N values from the queue at once,
+    /// values may not be meaningful if the queue is
+    /// not of length N.
+    ///
+    #[inline(always)]
+    pub fn peak_bitcounts<const M: usize>(&mut self) -> [usize; M] {
+        let mut values: [usize; M] = [0; M];
+        for i in 0..M {
+            let index = (self.front + i) % 16;
+            unsafe {
+                *values.get_unchecked_mut(i) = self.count_at(index);
+            }
+        }
+        values
+    }
+
+    ///
+    /// Internal use accessor to an initialized value.
+    ///
+    /// # Safety
+    /// This function is unsafe because it assumes that
+    /// the index is inbounds and initialized.
+    ///
+    unsafe fn value_at(&self, index: usize) -> usize {
+        *self.zigzag.get_unchecked(index)
+    }
+
+    ///
+    /// Internal use accessor to an initialized value.
+    ///
+    /// # Safety
+    /// This function is unsafe because it assumes that
+    /// the index is inbounds and initialized.
+    ///
+    unsafe fn count_at(&self, index: usize) -> usize {
+        *self.bitcount.get_unchecked(index)
+    }
+
+    ///
+    /// Internal use mutable accessor to an initialized value.
+    ///
+    /// # Safety
+    /// This function is unsafe because it assumes that
+    /// the index is inbounds. It does *not* drop the value
+    /// at the index if it was initialized; therefore, T must
+    /// be trivially dropable.
+    ///
+    unsafe fn write<T: Bits + Sized>(&mut self, index: usize, t: T) {
+        let (zbits, zcount) = t.zigzag_bits();
+        *self.zigzag.get_unchecked_mut(index) = zbits;
+        *self.bitcount.get_unchecked_mut(index) = zcount;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn can_init() {
+        let queue: CompressionQueue<10> = CompressionQueue::new();
+        assert_eq!(queue.len(), 0);
+        assert_eq!(queue.is_full(), false);
+        assert_eq!(queue.is_empty(), true);
+    }
+
+    // #[test]
```

Comment:
Should these tests be disabled? Some of this functionality looks like it does not need to be tested.

Reply:
Enabled and updated the unit tests according to CompressionQueue in commit [e90280e](https://github.com/qsib-cbie/tsz/pull/45/commits/e90280ec972c8e5ec2fc98867dcd24b04841be23).